sound/soc/codecs/rt715-sdca.c

```diff
@@ -88,171 +88,338 @@ static int rt715_sdca_set_amp_gain_put(struct snd_kcontrol *kcontrol,
+			gain_val[i * 2] = boost_max;
+		val_l = gain_val[i * 2];
+		gain_val[i * 2] = (gain_val[i * 2] * 10) << gain_sft;
+		/* R channel */
+		gain_val[i * 2 + 1] = ucontrol->value.integer.value[i * 2 + 1];
+		if (gain_val[i * 2 + 1] > boost_max)
+			gain_val[i * 2 + 1] = boost_max;
+		val_r = gain_val[i * 2 + 1];
+		gain_val[i * 2 + 1] = (gain_val[i * 2 + 1] * 10) << gain_sft;
+	}
+
+	for (i = 0; i < loop_cnt * 2; i++) {
+		if (i < 7) {
+			ret = regmap_write(rt715->mbq_regmap, reg_base + i,
+						gain_val[i]);
+			if (ret != 0) {
+				dev_err(component->dev, "Failed to write 0x%x=0x%x\n",
+					reg_base + i, gain_val[i]);
+				return ret;
+			}
+		} else {
+			ret = regmap_write(rt715->mbq_regmap, (reg_base - 1) | BIT(15),
+						gain_val[i]);
+			if (ret != 0) {
+				dev_err(component->dev, "Failed to write 0x%x=0x%x\n",
+					(unsigned int)((reg_base - 1) | BIT(15)), gain_val[i]);
+				return ret;
+			}
+		}
+	}
+
+	return 0;
+}
+
 static int rt715_sdca_set_amp_gain_get(struct snd_kcontrol *kcontrol,
 		struct snd_ctl_elem_value *ucontrol)
 {
 	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
 	struct soc_mixer_control *mc =
 		(struct soc_mixer_control *)kcontrol->private_value;
 	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
-	unsigned int val_l, val_r, ctl_l, ctl_r, neg_flag = 0;
+	unsigned int val_l, val_r, neg_flag = 0, ctl_val[2] = {0};
+	unsigned int i;
 	int ret;
 
 	ret = regmap_read(rt715->mbq_regmap, mc->reg, &val_l);
 	if (ret < 0)
-		dev_err(component->dev, "Failed to read 0x%x, ret=%d\n", mc->reg, ret);
+		dev_err(component->dev, "Failed to read 0x%x, ret=%d\n", mc->reg,
+				ret);
 	ret = regmap_read(rt715->mbq_regmap, mc->rreg, &val_r);
 	if (ret < 0)
 		dev_err(component->dev, "Failed to read 0x%x, ret=%d\n", mc->rreg,
 				ret);
-
 	/* L channel */
-	if (mc->shift == 8) {
-		ctl_l = (val_l >> mc->shift) / 10;
-	} else {
-		ctl_l = val_l;
-		if (ctl_l & BIT(15)) {
-			ctl_l = ~(val_l - 1) & 0xffff;
-			neg_flag = 1;
-		}
-		ctl_l *= 1000;
-		ctl_l >>= 8;
-		if (neg_flag)
-			ctl_l = mc->shift - ctl_l / RT715_SDCA_DB_STEP;
-		else
-			ctl_l = mc->shift + ctl_l / RT715_SDCA_DB_STEP;
+	ctl_val[0] = val_l;
+	if (ctl_val[0] & BIT(15)) {
+		ctl_val[0] = ~(val_l - 1) & 0xffff;
+		neg_flag = 1;
 	}
+	ctl_val[0] *= 1000;
+	ctl_val[0] >>= 8;
+	if (neg_flag)
+		ctl_val[0] = mc->shift - ctl_val[0]
+			/ RT715_SDCA_DB_STEP;
+	else
+		ctl_val[0] = mc->shift + ctl_val[0]
+			/ RT715_SDCA_DB_STEP;
 
 	neg_flag = 0;
 	/* R channel */
-	if (mc->shift == 8) {
-		ctl_r = (val_r >> mc->shift) / 10;
-	} else {
-		ctl_r = val_r;
-		if (ctl_r & BIT(15)) {
-			ctl_r = ~(val_r - 1) & 0xffff;
+	ctl_val[1] = val_r;
+	if (ctl_val[1] & BIT(15)) {
+		ctl_val[1] = ~(val_r - 1) & 0xffff;
+		neg_flag = 1;
+	}
+	ctl_val[1] *= 1000;
+	ctl_val[1] >>= 8;
+	if (neg_flag)
+		ctl_val[1] = mc->shift - ctl_val[1]
+			/ RT715_SDCA_DB_STEP;
+	else
+		ctl_val[1] = mc->shift + ctl_val[1]
+			/ RT715_SDCA_DB_STEP;
```

Comment:
ditto, best merge with the loop below. More cases below should be converted to loops too IMHO